File: Chapter 03/C#/Section_3_2_Coupling_Exercise/Before/UserDataService.cs
using System;
using System.Collections.Generic;

namespace Chapter03.CouplingTest.Before;

/// <summary>
/// A service that provides user data.
/// This version demonstrates a "chatty," fine-grained API that leads to high coupling.
/// </summary>
public class UserDataService
{
    // PROBLEM: This service's API is too fine-grained.
    // It forces any client that wants to build a user report to know exactly
    // which methods to call and in what order. The client has to do the work
    // of orchestrating the data retrieval. This exposes the internal structure
    // of the data and creates a tight coupling with any client that uses it.
    
    public string GetUserName(int userId) 
    {
        Console.WriteLine("    [Service] Fetching Name...");
        return "Jane Doe"; 
    }
    
    public string GetUserEmail(int userId) 
    { 
        Console.WriteLine("    [Service] Fetching Email...");
        return "jane.doe@example.com"; 
    }
    
    public List<string> GetUserOrderIds(int userId) 
    { 
        Console.WriteLine("    [Service] Fetching Order IDs...");
        return new List<string> { "A123", "B456" }; 
    }
    
    public decimal GetOrderTotal(string orderId) 
    { 
        Console.WriteLine($"    [Service] Fetching Total for Order {orderId}...");
        return 99.95m; 
    }
}

File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_2_coupling_exercise/before/UserDataService.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_2_coupling_exercise.before;

import java.util.Arrays;
import java.util.List;

/**
 * ANTI-PATTERN: The "Chatty" API and Tight Coupling.
 * * ARCHITECTURE PROBLEM: This service exposes highly granular methods. While this 
 * might seem like it promotes reuse, it forces the client to make multiple 
 * sequential calls to assemble a complete picture of a User.
 * * If this were a remote microservice or a network boundary, every single 
 * method call would incur latency. Furthermore, the client is forced to know 
 * too much about how to piece this data together.
 */
public class UserDataService {
    
    // 🚨 ARCHITECTURE WARNING: Forces the client to make a separate call just for the name.
    public String getUserName(int userId) {
        System.out.println("    [Service] Fetching Name...");
        return "Jane Doe";
    }

    // 🚨 ARCHITECTURE WARNING: Forces the client to make a separate call just for the email.
    public String getUserEmail(int userId) {
        System.out.println("    [Service] Fetching Email...");
        return "jane.doe@example.com";
    }

    public List<String> getUserOrderIds(int userId) {
        System.out.println("    [Service] Fetching Order IDs...");
        return Arrays.asList("A123", "B456");
    }

    /**
     * 🚨 ARCHITECTURE WARNING: High Coupling to Data Structure.
     * By forcing the client to fetch the total for each order individually, 
     * the service delegates its domain responsibility (calculating a user's total) 
     * to the client.
     */
    public double getOrderTotal(String orderId) {
        System.out.println("    [Service] Fetching Total for Order " + orderId + "...");
        return 99.95;
    }
}